File: bowershub-ai/frontend/src/services/financeTransactions.ts
/**
 * Typed client for the transactions explorer (routers/finance_transactions.py).
 * No `any` at the boundary.
 */
import { api } from './api'

export interface TxnRow {
  id: string
  posted_date: string
  description: string | null
  merchant_key: string | null
  amount: number
  account_id: string | null
  account_name: string | null
  category_id: number | null
  category_name: string | null
  is_transfer: boolean
  is_investment: boolean
  is_split: boolean
  cleared: boolean
}

export interface Subtotal { category: string; total: number; count: number }
export interface Totals { income: number; spending: number }

export interface TxnSearchResult {
  items: TxnRow[]
  count: number
  subtotals: Subtotal[]
  totals: Totals
}

export type TxnStatus = 'all' | 'uncategorized' | 'spending' | 'income' | 'transfers'

export interface TxnQuery {
  q?: string
  category_id?: number
  month?: string          // YYYY-MM-01
  start?: string          // YYYY-MM-DD (inclusive)
  end?: string            // YYYY-MM-DD (inclusive)
  status?: TxnStatus
  sort?: 'date' | 'amount' | 'category' | 'description'
  order?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export const financeTransactions = {
  async search(query: TxnQuery): Promise<TxnSearchResult> {
    const p = new URLSearchParams()
    if (query.q) p.set('q', query.q)
    if (query.category_id != null) p.set('category_id', String(query.category_id))
    if (query.month) p.set('month', query.month)
    if (query.start) p.set('start', query.start)
    if (query.end) p.set('end', query.end)
    if (query.status && query.status !== 'all') p.set('status', query.status)
    if (query.sort) p.set('sort', query.sort)
    if (query.order) p.set('order', query.order)
    if (query.limit != null) p.set('limit', String(query.limit))
    if (query.offset != null) p.set('offset', String(query.offset))
    const { data } = await api.get(`/api/finance/transactions?${p.toString()}`)
    return data as TxnSearchResult
  },
}

File: bowershub-ai/backend/routers/finance_transactions.py
"""Transactions explorer API (Monarch/Origin-style). A single flexible read
endpoint backing the unified Finance → Transactions view: filter (text/category/
month/account/status), sort, paginate, with allocation-aware subtotals + totals.
Read-only; auth via get_current_user; DB-down → typed 503.
"""

from __future__ import annotations

import logging
from datetime import date
from typing import Optional

import asyncpg
from fastapi import APIRouter, Depends, HTTPException

from backend.database import get_pool
from backend.middleware.auth import get_current_user
from backend.services.transactions_query import search_transactions

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/finance", tags=["finance-transactions"])

_STATUSES = {"all", "uncategorized", "spending", "income", "transfers"}


@router.get("/transactions")
async def list_transactions(
    q: Optional[str] = None,
    category_id: Optional[int] = None,
    month: Optional[date] = None,
    start: Optional[date] = None,
    end: Optional[date] = None,
    account_id: Optional[str] = None,
    owner: Optional[str] = None,
    status: str = "all",
    sort: str = "date",
    order: str = "desc",
    limit: int = 100,
    offset: int = 0,
    user: dict = Depends(get_current_user),
) -> dict:
    if status not in _STATUSES:
        raise HTTPException(status_code=400, detail=f"invalid status; one of {sorted(_STATUSES)}")
    try:
        async with get_pool().acquire() as conn:
            return await search_transactions(
                conn, q=q, category_id=category_id,
                month=month.replace(day=1) if month else None,
                start=start, end=end,
                account_id=account_id, owner=owner, status=status,
                sort=sort, order=order,
                limit=limit, offset=offset)
    except (asyncpg.PostgresError, OSError, RuntimeError) as e:
        logger.warning("finance_transactions: DB unavailable: %s", e)
        raise HTTPException(status_code=503, detail="Finance database unavailable.")
